Automattic/wp-calypso
# Limit Global Styles: Refactor the Style Selection implementation

#### Proposed Changes

* Refactor the implementation of the Limited Global Styles flows within the Style Selection in Onboarding.
* Update the design picker primary CTA to also include an "Unlock this style" option for premium Global Styles.
* The "Unlock this style" button opens a new modal, prompting to either "Upgrade Plan" or "Try it out first".
  * "Try it out first" simply pick the selected theme and style variation and proceed to set up the site.
  * "Upgrade plan" navigates to Checkout for a Premium plan. On purchase completion, the user will be redirected back to the Style Selection with both theme and style variation preselected.

| Before | After |
| - | - |
| <img width="1515" alt="Screenshot 2022-12-02 at 17 26 18" src="https://user-images.githubusercontent.com/2070010/205350283-e70c3cc3-1fbe-44a3-aa61-867892e29c32.png"> | <img width="1513" alt="Screenshot 2022-12-02 at 17 25 09" src="https://user-images.githubusercontent.com/2070010/205350309-8330c843-ae2c-4960-b49a-0f0d42ff20f5.png"><br><img width="789" alt="Screenshot 2022-12-06 at 12 18 15" src="https://user-images.githubusercontent.com/2070010/205910677-45e26bda-ada3-4aac-9aa5-c10c560ffef6.png"> |

##### Technical Changes

* Move the primary button into its own `getPrimaryActionButton` function to make the various conditions clearer (also it would have been a nested ternary otherwise 🙂 ).
* Basically copy-paste the `UpgradeModal` into `PremiumGlobalStylesModal`, a bit simpler and without any of the beta/feature-flagged logic.
  * I avoided creating a shared solution because I'm unfamiliar with the area, and a simple modal is not worth risking billions of regressions.
* Extend the `theme` query param preselection logic to style variations as well (through the `?style` param, but I'm open to different names if this feels too generic).
* Add two new Tracks events that need to be registered:
  * `calypso_signup_design_premium_global_styles_modal_show`
  * `calypso_signup_design_premium_global_styles_modal_checkout_button_click`

#### Testing Instructions

* Create a new free site and follow the steps until Style Selection.
* Apply the `wpcom-limit-global-styles` sticker to the newly created site and refresh the page.
* Choose a free theme without style variations.
  * ☑️ Ensure the CTA is "Continue" (unchanged).
* Choose a premium theme with style variations.
  * Try selecting the default or premium variations.
  * ☑️ Ensure the CTA is "Unlock theme" in all cases (unchanged).
* Choose a free theme with style variations.
  * Pick the default variation.
  * ☑️ Ensure the CTA is "Continue" (unchanged).
  * Pick a non-default ("premium") variation.
  * 🆕 Ensure the CTA is "Unlock this style".
  * Click on the CTA and observe the new modal.
  * ➡️ Compare the content with the one proposed in [the issue](https://github.com/Automattic/dotcom-forge/issues/1225) (ignore the screenshot). It should be **roughly** the same, but I've done some little changes to reuse existing strings.
  * Click on "Try it out first" and let the site setup complete.
  * Check the site in the Site Editor.
  * ➡️ It should use the selected ("premium") style variation.
  * Check the site on the front end.
  * ➡️ It should use the default style variation.
    * Yes, I'm aware this UX is not excellent; let's do it one step at a time.
  * Go back to the Style Selection in onboarding. It should work even if the site is complete; otherwise, restart the test from the first step. 😅 
  * Select a free theme with style variations, then choose a non-default ("premium") variation, and click on "Unlock this style" to open the modal.
  * This time, select "Upgrade plan".
  * ➡️ You should navigate to the Checkout page with a Premium plan in the cart.
  * Complete the purchase.
  * ➡️ Ensure you are redirected to the Style Selection, with the correct theme and style variation preselected.
  * ☑️ Now that you are on Premium, all mentions of "premium" variations should be gone (unchanged).

#### Pre-merge Checklist





Fixes https://github.com/Automattic/dotcom-forge/issues/1225
